core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,246 @@
+                indexNameExpressionResolver, request, request, executor);
+    }
+
+    /**
+     * Called on the primary with a reference to the {@linkplain IndexShard} to modify.
+     */
+    protected abstract WriteResult<Response> onPrimaryShard(Request request, IndexShard indexShard) throws Exception;
+
+    /**
+     * Called once per replica with a reference to the {@linkplain IndexShard} to modify.
+     *
+     * @return the translog location of the {@linkplain IndexShard} after the write was completed or null if no write occurred
+     */
+    protected abstract Translog.Location onReplicaShard(Request request, IndexShard indexShard);
+
+    @Override
+    protected final WritePrimaryResult shardOperationOnPrimary(Request request) throws Exception {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        WriteResult<Response> result = onPrimaryShard(request, indexShard);
+        return new WritePrimaryResult(request, result.getResponse(), result.getLocation(), indexShard);
+    }
+
+    @Override
+    protected final WriteReplicaResult shardOperationOnReplica(Request request) {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        Translog.Location location = onReplicaShard(request, indexShard);
+        return new WriteReplicaResult(indexShard, request, location);
+    }
+
+    /**
+     * Simple result from a write action. Write actions have static method to return these so they can integrate with bulk.
+     */
+    public static class WriteResult<Response extends ReplicationResponse> {
+        private final Response response;
+        private final Translog.Location location;
+
+        public WriteResult(Response response, @Nullable Location location) {
+            this.response = response;
+            this.location = location;
+        }
+
+        public Response getResponse() {
+            return response;
+        }
+
+        public Translog.Location getLocation() {
+            return location;
+        }
+    }
+
+    private class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
+        private final ReplicatedWriteRequest<?> request;
+        volatile boolean finishedWrite;
+        volatile ActionListener<Response> listener = null;
+
+        public WritePrimaryResult(Request request, Response finalResponse,
+                                  @Nullable Translog.Location location,
+                                  IndexShard indexShard) {
+            super(request, finalResponse);
+            this.request = request;
+            finishWrite(indexShard, request, location);
```

Comment:
add a comment here that we do it before replication so refresh listeners will go concurrently with waiting for replicas?